FIP-0049 (actor events): align with implementation. (#581)
filecoin-project/FIPs: FIPS/fip-0049.md

---
fip: "0049"

> title: Actor events

Reviewer: ```suggestion
title: Actor Events & Logs
```

Author: Is this a personal style comment or is there a rule that I've overlooked? I would keep as-is unless there's clear justification. The recommended style in scientific writing is lower case nouns for theories, concepts, etc.

Reviewer: Nope - not personal style. It's a pattern that past FIPs have been used + FIP-0001 does the same https://github.com/filecoin-project/FIPs/tree/master/FIPS

Author: Happy to e=>E if it gets this merged...

Reviewer: lolol

Reviewer: hahah

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Simple Summary](#simple-summary)
- [Abstract](#abstract)
- [Change Motivation](#change-motivation)
- [Specification](#specification)
  - [New chain types](#new-chain-types)
  - [Chain commitment](#chain-commitment)
  - [New `vm::emit_event` syscall](#new-vmemit_event-syscall)
  - [Gas costs](#gas-costs)
  - [Mapping to EVM logs](#mapping-to-evm-logs)
- [Design Rationale](#design-rationale)
- [Backwards Compatibility](#backwards-compatibility)
- [Test Cases](#test-cases)
- [Security Considerations](#security-considerations)
- [Incentive Considerations](#incentive-considerations)
- [Product Considerations](#product-considerations)
  - [Required for EVM compatibility](#required-for-evm-compatibility)
  - [Built-in actors observability](#built-in-actors-observability)
- [Implementation](#implementation)
  - [Reference implementation](#reference-implementation)
  - [Client handling](#client-handling)
- [Copyright](#copyright)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Simple Summary

> This FIP adds the ability for actors to emit externally observable events during

Reviewer: ```suggestion
This FIP adds the ability for actors to emit externally observable onchain events during
```

Reviewer: it is not clear to me what is "external" with the current description

> execution. Events are fire-and-forget, and can be thought of as a side effect

Reviewer: im not really getting the "side effect" analogy choice - can you elaborate more on why you chose this word? 

(wydt about "asynchronous state transition signal/trigger"?)

> from execution. The payloads of events are self-describing objects signalling

Reviewer: ```suggestion
from execution. The payloads of events are self-describing objects signaling
```

Author: North Americans... :D

> that some relevant circumstance, action or transition has ocurred. Actor events

Reviewer: ```suggestion
that some relevant state transition has occurred. Actor events
```

Author: Can you please explain your suggestion here, @jennijuju? (Generally, for all other comments, please explain your suggestion unless it's a typo). We **do not** want to limit actor events to state transitions only. For example, a forbidden action could not result in a state transition, but still be recorded as an event (e.g. `TokenTransferDenied`).

Reviewer: AFAIK, (in etherum), any actions that cause a failed message execution will lead to all state changes to be reverted, including the events. (quick google verification, [here](https://ethereum.stackexchange.com/questions/90535/will-event-be-emitted-on-failed-transaction))

> We do not want to limit actor events to state transitions only. For example, a forbidden action could not result in a state transition,

In this example, according to the definition below: _When an actor exits abnormally (exit code > 0), events emitted from that invocation and its transitive callee are discarded._ Could you please provide an example that a forbidden action in actors that will result in an exit code =0 (successful execution?)

Reviewer: Still seeking for an explanation.

enable external agents to observe the activity on chain, and may eventually

> become the source of internal message triggers.

Author: Note: "Internal transaction" is an imprecise term and I'm not aware of us using it at all (if it's used somewhere, it should be revised). Rationale: a transaction is a single unit of execution, which in our system can only be the external (on-chain) message. 

An internal message is a send from one actor to another.

> become the source of internal message triggers.

Author: This is a speculative feature, and this means that an actor event could lead to a send.

> become the source of internal message triggers.

Reviewer: > An internal message is a send from one actor to another.

\ I think we use "internal invocation" more often but its not specifically defined, internal messages also make sense to me

   approaches to obtain the data they needed, e.g. forking and instrumenting
   built-in actors code, or replaying messages and performing state tree diffs.
   
   While this may have been tolerable while the system had a limited number of
   actors, such approaches will no longer scale in a world of discretionary
   programs on chain.
   
   A future FIP could enhance built-in actors to emit events when power changes,
   sectors are onboarded, deals are made, sectors are terminated, penalties are
   incurred, etc.

> 2. The upcoming introduction of the Filecoin Ethereum Virtual machine runtime

Reviewer: ```suggestion
2. The upcoming introduction of the Filecoin Ethereum Virtual Machine(EVM) runtime
```

> 2. The upcoming introduction of the Filecoin Ethereum Virtual machine runtime

Reviewer: link to the FEVM fip plz

Author: Once it's merged as a draft, I will post it here. I won't link to a PR.

   necessitates an event mechanism at the protocol layer to handle the
   `LOG{0..4}` EVM opcodes, and the subsequent serving of such data through

>    JSON-RPC methods.

Reviewer: ```suggestion
   JSON-RPC APIs to application/tooling developers.
```

Author: I suspect you are intending to add more colour here.

The problem is that JSON-RPC APIs are not serving data to _developers_ (the people), but to external agents, so this is imprecise.

I will add the "external" keyword, that's sufficient.

Author: (Although JSON-RPC implies external...)

>    JSON-RPC methods.

Reviewer: ```suggestion
   JSON-RPC APIs to application/tooling developers.
```

Author: Duplicated comment.

## Specification

### New chain types

> We introduce a DAG-CBOR encoded `StampedEvent` type to represent an actor event

Reviewer: ```suggestion
We introduce a[ DAG-CBOR](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md) encoded `StampedEvent` type to represent an actor event
```

Author: Spacing in the suggestion is wrong, so I can't accept it as-is. I'm also using the lookup link format, so I will add it to the bottom.

> as committed on chain.

Reviewer: ```suggestion
as emitted on chain.
```

Author: No, that's incorrect.

Reviewer: elaborate why please? what is an committed event and whats an emitted event?

```rust
/// Represents an event emitted throughout message execution, stamped by the FVM
/// with additional metadata.
struct StampedEvent {

>     /// Carries the actor ID of the emitting actor (stamped by the FVM).

Author: No, that's incorrect. This carries the actor ID, not the actor's ID address.

>     /// Carries the actor ID of the emitting actor (stamped by the FVM).

Reviewer: Can you specify what actor ID is then?  I cannot find any definition in [filecoin spec](https://spec.filecoin.io) or in any past FIPs. The spec only have mentioned `actor id address`

>     /// Carries the actor ID of the emitting actor (stamped by the FVM).

Author: The actor ID is the internal identifier of the actor in the state tree, and the payload of the ID address. If it's not specified in the spec, it needs to be fixed there. I can add a footnote if others share this concern, but anybody reading this spec should know what an actor ID is (prerequired knowledge).

>     /// Carries the actor ID of the emitting actor (stamped by the FVM).

Reviewer: hrm, Im also not able to find any reference in spec-actor or go-state-types about actor ID. Ive been working on Filecoin for two years, tho it's certain that I still have unknown knowledge of this complicated system!  @raulk or other @filecoin-project/fips-editors , can one of you help me find a reference or explanation on this plz?

>     /// Carries the actor ID of the emitting actor (stamped by the FVM).

Reviewer: @raulk and I did a walk through here and found out that the we actually store ID address in the state trees for most filecoin things. The exception here is the recently added datacap actor. And we use ID addresses  in the clients - thats where the confusion was coming from. chatting with aayush and steb, we think it is worth considering to do a migration and use `ActorID` in places.

    emitter: u64,
    /// Payload of the event, as emitted by the actor.
    event: ActorEvent,
}

/// An event as emitted by the actor.
struct ActorEvent {
    /// The key-value entries that make up the event.
    entries: [Entry],
}

struct Entry {
    /// A bitmap conveying metadata or hints about this entry.

>     flags: u64,

Author: No, this value carries multiple flags in the form of bitflags/bitmask.

## Specification

### New chain types

We introduce the following data model to represent and commit events on chain.
These objects are DAG-CBOR encoded with tuple encoding.

```rust
/// Represents an event emitted throughout message execution, stamped by the FVM
/// with additional metadata, and committed on chain.
struct StampedEvent {
    /// Carries the actor ID of the emitting actor (stamped by the FVM).
    emitter: u64,
    /// Payload of the event, as emitted by the actor.
    event: ActorEvent,
}

/// An event as originally emitted by the actor.
struct ActorEvent {
    /// The key-value entries that make up the event.
    entries: [Entry],
}

struct Entry {
    /// A bitmap conveying metadata or hints about this entry.
    flags: u64,
    /// The key of this entry.
    key: String,
    /// The value of this entry.
    /// Any DAG-CBOR encoded type, but currently restricted to only byte strings (CBOR major type 2).
    value: any,
}
```

Its main constituent is a list of ordered key-value **entries**. This approach
is inspired by structured logging concepts. **Keys** are UTF-8 strings, while
**values** can be any DAG-CBOR encodeable type.

> However, this FIP restricts accepted types to byte strings only (CBOR major type

Reviewer 2: I had to jump to go back to the sentence above to understand what `types` refers to. Maybe consider changing to "types of values" or "values' types"?

Author: Thanks, will adjust the prose.

Author: Fixed in 05f693f.

supporting on-chain event subscriptions, special type indicators (e.g. "value is
an address"), and more.

Note: DAG-CBOR encoding guarantees that flag bitmaps with up to 4 bits will be
minimally encoded as a single byte (major type 0, with inlined value in
additional data).

Choosing a _list of tuples_ instead of a map representation is deliberate, as it
enables undistinguished repeatable keys, and, thus, array-like entries, e.g.
signers in a multisig transaction.

Over time, we expect the community to standardise on a set of keys through FRC
proposals.

**Event examples**

The following examples demonstrate how to use these structures in practice.
Payloads are DAG-CBOR encoded types.

_Ethereum-style event_

See FIP-0054 for more information.

```rust
Event {
    emitter: 1234,
    entries: [
        (0x03, "t1", <32-byte array>),  // CBOR major type 2
        (0x03, "t2", <32-byte array>),  // CBOR major type 2
        (0x03, "t3", <32-byte array>),  // CBOR major type 2
        (0x03, "t4", <32-byte array>),  // CBOR major type 2
        (0x03, "p", <32-byte array>),   // CBOR major type 2
    ],
}
```

_Fungible token transfer_

> **NOTE: value types not supported yet.**

Reviewer 3: ```suggestion
**NOTE: value types other than byte string are not supported yet.**
```

And ditto below

Author: Fixed in d002339.

```rust
// type, sender, and receiver indexed by key and value.
Event {
    emitter: 1260,
    entries: [
        (0x03, "type", "transfer"),     // CBOR major type 3
        (0x03, "sender", <actor id>),   // CBOR major type 0
        (0x03, "receiver", <actor id>), // CBOR major type 0
        (0x00, "amount", <token amount in atto precision>),  // CBOR major type 2, BE-encoded BigInt, Not indexed.
    ],
}
```

_Non-fungible token transfer_

**NOTE: value types not supported yet.**

```rust
// All entries indexed, carrying the event type as a key (transfer).
Event {
    emitter: 101,
    entries: [

>         (0x01, "transfer", null),

Reviewer 3: Nit: why is this inconsistent with the fungible token transfer using ("type", "transfer")?

Author: I added this example with a different key-value style to demonstrate a usage of the "index by key" flag. Since user code will pay per byte, I expect devs will come up with ways to produce compact event representations without information loss. On an abstract level, this follows this pattern for a schema:

```json
{ "transfer": { "sender": 1111, "receiver": 2222, "token_id": "0x12345678" }  }
```

As opposed to:

```json
{ "type": "transfer", "sender": 1111, "receiver": 2222, "token_id": "0x12345678" } 
```

  discarded.

At the end of successful message execution, the FVM serializes all retained
`StampedEvent`s to DAG-CBOR and aggregates them into an AMT with bitwidth 5, in
their natural position. The AMT's root CID is then returned to the client, along
with the events themselves, together with the other results from message
execution.

### Chain commitment

The existing `MessageReceipt` chain data structure is augmented with a new
`events_root` field:

```rust
struct MessageReceipt {
    exit_code: ExitCode,
    return_data: RawBytes,
    gas_used: i64,
    // Root of AMT<StampedEvent, bitwidth=5>
    // Absent when no events have been emitted.
    events_root: Option<Cid>,
}
```

Because the network agrees on the content of receipts through the
`ParentMessageReceipts` field on the `BlockHeader`, actor events are implicitly
committed to the chain via this new `events_root` field.

Note that because the FVM does not write receipts to the blockstore, neither
does it write the events AMT themselves. Doing so would result in a dangling
write. Clients are responsible for managing event data as they wish, just like

> they're responsible for managing receipts.

Reviewer 3: The correctness of this is dubious. The VM might not write, but the node probably does, or certainly could, in order to support subsequent queries. I think this paragraph could just be dropped without loss.

Author: Which part is incorrect? The FVM certainly does not write receipts, so writing the event AMTs would lead to a dangling write since they would be unreferenced unless the client writes the receipt.

Author: Basically what this is saying is: "look client, you're responsible for managing receipts, and that includes events."

is inspired by structured logging concepts. **Keys** are strings, while
**values** can be any DAG-CBOR encodeable type.

Every **entry** contains a u64 **flags** bitmap, which conveys metadata or hints
about the entry. These are the currently supported flags:

| hex    | binary      | meaning          |
|--------|-------------|------------------|
| `0x01` | `b00000001` | indexed by key   |

> | `0x02` | `b00000010` | indexed by value |

Author: Added.

Flags are additive, such that value 0x03 (`b00000011`) indicates that the entry
must be indexed by both key and value.

These flags are only used as a client hint. In the future, this flag and others
may lead to consensus-relevant outcomes, such as populating (adaptable) bloom
filters, committing block-level indices, supporting on-chain event
subscriptions, special type indicators (e.g. "value is an address"), and more.

Note: DAG-CBOR encoding guarantees that flags with up to 4 bits will be
minimally encoded as a single byte (major type 0, with inlined value in
additional data).

Choosing a _list of tuples_ instead of a map representation is deliberate, as it
enables undistinguished repeatable keys, and, thus, array-like entries, e.g.
signers in a multisig transaction.

> Over time, we expect the community to standardise on a set of keys through FRC

Author: The keys. There's an example below in _Multisig approval_.

proposals.

> **Event examples**

Reviewer: lets add an example where the value is indexed?

Author: Hm, can do.

The following examples demonstrate how to use these structures in practice.

> Payloads are DAG-CBOR encoded types.

Author: The spec? Or a link to the spec? I've added the latter, but I won't add the full spec here, it makes no sense.

> Payloads are DAG-CBOR encoded types.

Reviewer: yes link is fine.

_Fungible token transfer_

```rust
Event {

>     emitter: 1260,

Author: No, it's an actor ID, not an ID address.

>     emitter: 1260,

Reviewer: same question as above - not sure what an actor ID is, and how is an value assigned.

> is inspired by structured logging concepts. **Keys** are strings, while

Reviewer 4: ```suggestion
is inspired by structured logging concepts. **Keys** are utf-8 strings, while
```
All strings in rust are utf-8 (if we are sticking to the String types and not just bytes)

Author: UTF-8 encoding is implied by CBOR.

Author: Will add if it makes you happy (but with the right capitalization, i.e. UTF-8, since we're in heavy bikeshedding territory)

**values** can be any DAG-CBOR encodeable type.

Every **entry** contains a u64 **flags** bitmap, which conveys metadata or hints
about the entry. These are the currently supported flags:

| hex    | binary      | meaning          |
|--------|-------------|------------------|
| `0x01` | `b00000001` | indexed by key   |
| `0x02` | `b00000010` | indexed by value |

> Flags are additive, such that value 0x03 (`b00000011`) indicates that the entry

Reviewer 4: Is the additive property something we are enforcing for future flags as well or is it just the current state of things?

Author: Specifications are usually written perennially unless otherwise specified.

> must be indexed by both key and value.

Reviewer 4: Must? Below it describes flags as being client hints that _should_ be done.

Author: Will lower language.

A zero-valued bitflag indicates that no flags are in effect.

The flags introduced herein are only used as a client hint, in this case to
signal indexing intent. Clients should honor the request, and construct indices
that facilitate fast lookup.

In the future, this flag and others may lead to consensus-relevant outcomes,
such as populating (adaptable) bloom filters, committing block-level indices,
supporting on-chain event subscriptions, special type indicators (e.g. "value is
an address"), and more.

Note: DAG-CBOR encoding guarantees that flags with up to 4 bits will be
minimally encoded as a single byte (major type 0, with inlined value in
additional data).

Choosing a _list of tuples_ instead of a map representation is deliberate, as it
enables undistinguished repeatable keys, and, thus, array-like entries, e.g.
signers in a multisig transaction.

Over time, we expect the community to standardise on a set of keys through FRC
proposals.

**Event examples**

> The following examples demonstrate how to use these structures in practice.

Reviewer 4: ```suggestion
The following examples demonstrate these structures in practice.
```

Author: No, these examples demonstrate _usage_.

Payloads are DAG-CBOR encoded types.

_Fungible token transfer_

```rust

> // type, sender, receiver indexed.

Reviewer 4: ```suggestion
// type, sender, and receiver indexed by key and value.
```

Event {

>     emitter: 1260,

Reviewer 4: ```suggestion
    emitter: <actor id>,
```

Author: No, these are _concrete examples_ not templates.

    entries: [
        (0x03, "type", "transfer"),
        (0x03, "sender", <actor id>),
        (0x03, "receiver", <actor id>),

>         (0x00, "amount", <token amount in atto precision>),

Reviewer 4: ```suggestion
        (0x00, "amount", <token amount in atto precision>), // Not indexed.
```

    ],
}
```

_Non-fungible token transfer_

```rust
// All entries indexed, carrying the event type as a key (transfer).
Event {

>     emitter: 101,

Reviewer 4: ```suggestion
    emitter: <actor id>,
```

Author: I'm OK making payloads placeholders (as they are user defined), but not data stamped by the protocol.

    entries: [
        (0x01, "transfer", null),
        (0x03, "sender", <actor id>),
        (0x03, "receiver", <actor id>),
        (0x03, "token_id", <internal identifier>),
    ],
}
```

_Access control_

```rust
// Compact representation of an "access denied" event, indexing the object id but not its key.
Event {

>     emitter: 480,

Reviewer 4: ```suggestion
    emitter: <actor id>,
```

Author: I'm OK making payloads placeholders (as they are user defined), but not data stamped by the protocol.

    entries: [
        (0x01, "denied", null),
        (0x02, "id", <object id>),
    ],
}
```

_Multisig approval_

```rust

> // only type and signers indexed; signer is a variadic entry.

Reviewer 4: ```suggestion
// Only type and signers indexed; signer is a variadic entry.
```

Event {

>     emitter: 1678,

Reviewer 4: ```suggestion
    emitter: <actor id>,
```

```rust
Event {

>     emitter: 101,

Author: No.

>     entries: [  // all entries indexed, with type as key

Author: No, the type of the event is specified as a key.

Event {

>     emitter: 1678,

Author: No, see above.

}
```

> During message call stack execution, DAG-CBOR encoded `Event`s are accumulated

Reviewer: "call stack" as in including all internal tx as well?

Author: Yes.

within the FVM inside an AMT, preserving the insertion order. At the end of

> message execution, the AMT is flushed into the FVM's (buffered) blockstore, and

Reviewer: if the message execution is failed - does all the events accumulated so far got flushed?

Author: Good call, specified the behaviours on the various exit codes.

the root CID of the AMT is returned in the above `events` field.

Because the network agrees on the content of the receipts through the
`ParentMessageReceipts` on the `BlockHeader`, the events are implicitly

> committed to the chain state with no futher structural changes needed.

Reviewer: ```suggestion
emitted to the chain state with no further structural changes needed.
```

Author: No, `commit` is the right word here, `emit` changes the meaning to an inaccurate one. Thanks for the typo fix, I've fixed it.

> > TODO need to make syscall gas dynamic based on param length

Reviewer: ❗ todo is needed before moving to last call

## Incentive Considerations

No incentive considerations apply.

## Product Considerations

### Required for EVM compatibility

Even though we've been seeking to introduce on-chain observability for some time
now, we explicitly recognize that the immediate motivation is EVM compatibility.

### Built-in actors observability

Upcoming network versions should strive to update built-in actors to emit actor

> events for meaningful transitions and circumstances. This would allow

Reviewer: ```suggestion
events for meaningful state transitions. This would allow
```

Author: Rejected as per above comments.

> events for meaningful transitions and circumstances. This would allow

Reviewer: ```suggestion
events for meaningful state transitions. This would allow
```

Author: Rejected as per above comments.

observability, monitoring, accounting and other external tools to stop depending
on highly-coupled, internal and low-level mechanisms to source information, such
as state replays, state diffs, execution traces, and others.

> A project tackling this mission should begin by inventorizing all situations

Reviewer: Probably off-topic, tho for builtin actors' events, we shall think about who pays for events' emissions & why..

Author: These can only be paid for by gas. Gas is paid by the caller. I think the question you're hinting at is: "why would the caller subsidize the monitoring operations of third parties?" I don't think this is any different than the rationale of why callers pay for events in other chains. We're just being confronted with this question because we are introducing events after contracts already exist that would've benefitted from those events from day 0, so we have a sensation of optionality here.

that would emit events, designing the events, projecting the cost, and
prioritizing their implementation together with the community. There is likely a
sweet spot that strikes the optimal balance between value and cost.

## Implementation

### Reference implementation

The reference implementation is being carried out under
[`filecoin-project/ref-fvm`], and its first adoption is within the EVM runtime
actor under [`filecoin-project/builtin-actors`].

> ### Client handling

Reviewer: can you link a impl for indexing?

Author: It's not on master yet, so unsafe to link. I will add a PR link, but generally not happy doing that.

## Test Cases

Refer to events test cases in the [reference

> implementation](https://github.com/filecoin-project/ref-fvm/blob/master/testing/integration/tests/events_test.rs).

Reviewer 4: ```suggestion
implementation](https://github.com/filecoin-project/ref-fvm/blob/master/testing/integration/tests/events_test.rs). [permalink](https://github.com/filecoin-project/ref-fvm/blob/d8f5d86f28ecd03a9934240134dae83a9754f72e/testing/integration/tests/events_test.rs)
```

Attempts to write large amounts of data through events don't result in writes to
the underlying blockstore that backs the state tree. Thus, there is limited to
nil potential for contagion to other parts of the system in case of abuse.

## Incentive Considerations

No incentive considerations apply.

## Product Considerations

### Required for EVM compatibility

Even though we've been seeking to introduce on-chain observability for some time
now, we explicitly recognize that the immediate motivation is EVM compatibility.

### Observability of built-in actors

Upcoming network versions should strive to update built-in actors to emit actor
events for meaningful transitions and circumstances. This would allow
observability, monitoring, accounting and other external tools to stop depending
on highly-coupled, internal and low-level mechanisms to source information, such
as state replays, state diffs, execution traces, and others.

A project tackling this mission should begin by inventorizing all situations
that would emit events, designing the events, projecting the cost, and
prioritizing their implementation together with the community. There is likely a

> sweet spot that strikes the optimal balance between value and cost.

Reviewer 3: Nit: this paragraph is unnecessary here IMO

Author: I would like to retain some form of an consideration here on the value/cost balance, and the need for community discussion on what kinds of events would give the most bang for the buck.

- Events benefit observability tools.
- They do not benefit the message sender, who pays for the gas expenditures that those events incur. In a way, users are subsidizing the operating cost of observability tools.
- Built-in actors are of compulsory usage in the system. Users cannot choose alternative, cheaper implementations that could've emerged from open competition.

However, I can also move this thought to the discussion if it makes us feel better: https://github.com/filecoin-project/FIPs/discussions/484
